### PR TITLE
8311583: tableswitch broken by JDK-8310577

### PR DESCRIPTION
**Pull request by coleenp**
I demoted some types for -Wconversion warnings but they're supposed to be larger types to catch overflow for tableswitch and lookupswitch.  Restored old code (except int64_t not jlong).
Tested with tier1-4.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311583](https://bugs.openjdk.org/browse/JDK-8311583): tableswitch broken by JDK-8310577 (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to [63c28c0a](https://git.openjdk.org/jdk/pull/14788/files/63c28c0a234079c414c9a202006b8c2a416980cb)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14788/head:pull/14788` \
`$ git checkout pull/14788`

Update a local copy of the PR: \
`$ git checkout pull/14788` \
`$ git pull https://git.openjdk.org/jdk.git pull/14788/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14788`

View PR using the GUI difftool: \
`$ git pr show -t 14788`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14788.diff">https://git.openjdk.org/jdk/pull/14788.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14788#issuecomment-1624238826)

### CODE REVIEW

File: src/hotspot/share/interpreter/bytecodes.cpp [hunk 1]
```diff
@@ -385,12 +385,13 @@ int Bytecodes::special_length_at(Bytecodes::Code code, address bcp, address end)
       if (end != nullptr && aligned_bcp + 3*jintSize >= end) {
         return -1; // don't read past end of code buffer
       }
-      int lo = Bytes::get_Java_u4(aligned_bcp + 1*jintSize);
-      int hi = Bytes::get_Java_u4(aligned_bcp + 2*jintSize);
-      int len = (int)(aligned_bcp - bcp) + (3 + hi - lo + 1)*jintSize;
+      // Promote calculation to 64 bits to do range checks, used by the verifier.
+      int64_t lo = (int)Bytes::get_Java_u4(aligned_bcp + 1*jintSize);
+      int64_t hi = (int)Bytes::get_Java_u4(aligned_bcp + 2*jintSize);
```
Comment (dholmes-ora) [inline src/hotspot/share/interpreter/bytecodes.cpp:390]: Are the int casts needed?

File: test/hotspot/jtreg/runtime/verifier/TestTableSwitch.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ */
+
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.process.OutputAnalyzer;
+
+/*
+ * @test TestTableSwitch
+ * @bug 8311583
+ * @library /test/lib
+ * @compile tableswitchp1.jasm lookupswitchp1.jasm
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/verifier/TestTableSwitch.java:31]: Nit: the filenames and class names could use normal Java camel-case naming style.

File: test/hotspot/jtreg/runtime/verifier/tableswitchp1.jasm [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+public class tableswitchp1 version 50:0 {
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/verifier/tableswitchp1.jasm:24]: Please add a comment indicating what rule/condition this handwritten bytecode is breaking. I think it is just low > hi

File: test/hotspot/jtreg/runtime/verifier/lookupswitchp1.jasm [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+public class lookupswitchp1 version 50:0 {
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/verifier/lookupswitchp1.jasm:24]: Please add a comment indicating what rule/condition this handwritten bytecode is breaking.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Functional fix looks good. A couple of suggestions around the test.

Thanks.

File: src/hotspot/share/interpreter/bytecodes.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/interpreter/bytecodes.cpp:390]: Yes they are.    get_Java_u4 returns u4 (unsigned) and the 'int' cast sign extends it to int, then further sign extended to int64_t for the length calculation.  The tableswitch bytecode operands are signed int.

```
static inline u4   get_Java_u4(address p)           { return get_Java<u4>(p); }
```

File: test/hotspot/jtreg/runtime/verifier/tableswitchp1.jasm [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+public static Method runTable:"()V"
+	stack 4 locals 3
+{
+    iconst_0;
+    bytecode 170; // tableswitch
+    bytecode 0,0; // padding
+    bytecode 0,0,0,15; // default target == L16
+    bytecode 0x80,0,0,0; // 0x80000000 low
+    bytecode 0x7F,0xFF,0xFF,0xFF; // 0x7fffffff high
+    // no jump table, but gets past verifier
```
Comment (dean-long) [inline test/hotspot/jtreg/runtime/verifier/tableswitchp1.jasm]:
```suggestion
    // illegal empty jump table
```

**Review by dean-long — APPROVED** [review]
Looks good except for jasm comment I suggest improving.

File: src/hotspot/share/interpreter/bytecodes.cpp (hunk 1 above)
Comment (dean-long) [reply src/hotspot/share/interpreter/bytecodes.cpp:390]: I think this should really be a new API with a name like get_Java_s4(), but that can be a separate RFE.  There are other functions in globalDefinitions.hpp where we try really hard not to use undefined or implemention-defined behavior, and casting from unsigned to signed seems to be one of those, at least until c++20 fixes it.

**Review by dholmes-ora — APPROVED** [review]
Nothing further from me. Thanks

**Comment by coleenp** [comment]
Thank you for reviewing Dean and David, and for finding this, Dean.
/integrate

**Pull request closed without merging**